src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceLocator.cs

```diff
@@ -41,26 +38,25 @@ public static void InitializePackageServiceProvider(IAsyncServiceProvider provid
             }
         }
 
+        /// <inheritdoc cref="GetInstanceAsync{TService}"/>
         public static TService GetInstance<TService>() where TService : class
         {
             return NuGetUIThreadHelper.JoinableTaskFactory.Run(GetInstanceAsync<TService>);
         }
 
+        /// <summary>
+        /// Fetches a service that may be registered with the DTE, MEF or the VS service provider.
+        /// This method may switch to the UI thread.
+        /// </summary>
+        /// <typeparam name="TService"></typeparam>
+        /// <returns>The instance of the service request, <see langword="null"/> otherwise. </returns>
+        /// <remarks>
+        /// Prefer <see cref="GetComponentModelServiceAsync{TService}{TService}"/> over this requesting a MEF service.
+        /// A general rule is that only non-NuGet VS services should be retrieved this method.
+        /// </remarks>
         public static async Task<TService> GetInstanceAsync<TService>() where TService : class
         {
-            // VS Threading Rule #1
-            // Access to ServiceProvider and a lot of casts are performed in this method,
-            // and so this method can RPC into main thread. Switch to main thread explictly, since method has STA requirement
-            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
`GetInstanceAsync` makes 3 attempts, returning early when the instance is found.

1. DTE service
2. MEF service
3. VS Service

1 and 3 need the UI thread, and by removing this switch to UI thread, trying to get a VS service from a background thread will:

1. Call into `GetDTEServiceAsync` on the background thread
2. `GetDTEServiceAsync` switches to the UI thread
3. Whatever the return value, the `await` in `GetInstanceAsync` causes the continuation to be scheduled on a background thread
4. Calls `GetComponentModelServiceAsync`, which completes on the current (background) thread
5. Calls into `GetGlobalServiceAsync`, which switches to the UI thread
6. The `await` in `GetInstanceAsync` causes the continuation to be scheduled on a background thread
7. Finally return the value to the caller. I don't know the .NET async implementation well enough to know if the caller's continuation happens immediately, or is scheduled.

The point I'm trying to make is that all these thread switches cause latency while the task is scheduled, and then an appropriate thread, when not busy, can pick it up and run it. The more thread switches that happen, the more latency, especially on the UI thread, since there can only be one of those.

I don't know if it's a safe assumption that DTE services will never be available in MEF, but if that's true, you can re-arrange the order to try MEF first on the current thread, then switch to the UI thread before trying DTE and VS/global services, to minimize thread switching.  On the other hand, if a service is available both via DTE and MEF, it's probably better for us to get it via MEF and not need a thread switch at all.

Otherwise, I don't think we should change `GetInstanceAsync` to stop switching to the UI thread. Your other changes to have callers use `GetComponentModelService` instead of `GetInstanceAsync` is great, but I think those changes should be sufficient for this PR. This change to `GetInstanceAsync` can, in my opinion, be undone.